vk/middleware.go

```diff
@@ -1,72 +1,27 @@
 package vk
 
-import (
-	"net/http"
-)
-
-// Middleware represents a handler that runs on a request before reaching its handler
-type Middleware func(*http.Request, *Ctx) error
-
-// Afterware represents a handler that runs on a request after the handler has dealt with the request
-type Afterware func(*http.Request, *Ctx)
-
-// ContentTypeMiddleware allows the content-type to be set
-func ContentTypeMiddleware(contentType string) Middleware {
-	return func(r *http.Request, ctx *Ctx) error {
-		ctx.RespHeaders.Set(contentTypeHeaderKey, contentType)
-
-		return nil
-	}
-}
-
-// CORSMiddleware enables CORS with the given domain for a route
-// pass "*" to allow all domains, or empty string to allow none
-func CORSMiddleware(domain string) Middleware {
-	return func(r *http.Request, ctx *Ctx) error {
-		enableCors(ctx, domain)
-
-		return nil
-	}
-}
-
-// CORSHandler enables CORS for a route
-// pass "*" to allow all domains, or empty string to allow none
-func CORSHandler(domain string) HandlerFunc {
-	return func(r *http.Request, ctx *Ctx) (interface{}, error) {
-		enableCors(ctx, domain)
-
-		return nil, nil
-	}
-}
-
-func enableCors(ctx *Ctx, domain string) {
-	if domain != "" {
-		ctx.RespHeaders.Set("Access-Control-Allow-Origin", domain)
-		ctx.RespHeaders.Set("X-Requested-With", "XMLHttpRequest")
-		ctx.RespHeaders.Set("Access-Control-Allow-Headers", "Accept, Content-Type, Content-Length, Accept-Encoding, Authorization, cache-control")
-	}
-}
-
-// generate a HandlerFunc that passes the request through a set of Middleware first and Afterware after
-func augmentHandler(inner HandlerFunc, middleware []Middleware, afterware []Afterware) HandlerFunc {
-	return func(r *http.Request, ctx *Ctx) (interface{}, error) {
-		defer func() {
-			// run the afterware (which cannot affect the response)
-			// even if something in the request chain fails
-			for _, a := range afterware {
-				a(r, ctx)
-			}
-		}()
-
-		// run the middleware (which can error to stop progression)
-		for _, m := range middleware {
-			if err := m(r, ctx); err != nil {
-				return nil, err
-			}
+// Middleware type describes a handler that wraps another handler.
+type Middleware func(handlerFunc HandlerFunc) HandlerFunc
+
+// WrapMiddleware will take a slice of middlewares and a handler, and then wrap the handler into the middlewares with
+// the last middleware being the closest to the handler.
+//
+// mws := []Middleware{
+//    error,
+//    auth,
+//    cors,
+// }
+// h := WrapMiddleware(mws, myHandler)
+//
+// In this instance the flow is
+// Request -> error -> auth -> cors -> myHandler -> cors -> auth -> error -> Response
```

Comment:
This comment confused me for a moment. At first glance it looked like each middleware function was called twice, but that didn't make sense. So let me confirm my understanding as it seems slightly different (simpler) from the explanations in the PR description and comment.

```go
// psuedo go
response := error(req) {
    // error checking
    return auth(req) {
        // auth logic
        return cors(req) {
            // set cors headers
            return myHandler(req) { 
                // finally the actual handler
            }
        }
    }
}
```

Reply:
yep, that pretty much is what's happening. The before / after things look like this:
```go
// psuedo go
response := tracing(req) {
    // this is a before/after type
    span := tracing.StartSpan()
    defer span.Stop()

    resp, err := error(req) {
        // this would be an "after type", so nothing is happening here

        resp, err = auth(req) {
            // auth logic h appens before passing it to the next controller 

            return cors(req) {
                // set cors headers
                return myHandler(req) { 
                    // finally the actual handler
                }
            }
        }

        if err == nil {
            return resp, nil
        }

        // do the error checking here, turn the error into a 4xx or 5xx response instead
    }

    // defer runs here, after the error returned, to close tracing
}
```